File: app/src/main/java/com/elementary/tasks/navigation/settings/voice/TimeOfDayFragment.java
package com.elementary.tasks.navigation.settings.voice;

import android.os.Bundle;
import androidx.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.elementary.tasks.R;
import com.elementary.tasks.core.utils.TimeUtil;
import com.elementary.tasks.core.views.roboto.RoboTextView;
import com.elementary.tasks.databinding.FragmentTimeOfDayLayoutBinding;
import com.elementary.tasks.navigation.settings.BaseSettingsFragment;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class TimeOfDayFragment extends BaseSettingsFragment implements View.OnClickListener {

    private RoboTextView nightTime, eveningTime, dayTime, morningTime;
    private int morningHour, morningMinute;
    private int dayHour, dayMinute;
    private int eveningHour, eveningMinute;
    private int nightHour, nightMinute;
    private boolean is24;
    private SimpleDateFormat format = new SimpleDateFormat("HH:mm", Locale.getDefault());

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        FragmentTimeOfDayLayoutBinding binding = FragmentTimeOfDayLayoutBinding.inflate(inflater, container, false);
        nightTime = binding.nightTime;
        nightTime.setOnClickListener(this);
        eveningTime = binding.eveningTime;
        eveningTime.setOnClickListener(this);
        dayTime = binding.dayTime;
        dayTime.setOnClickListener(this);
        morningTime = binding.morningTime;
        morningTime.setOnClickListener(this);

        is24 = getPrefs().is24HourFormatEnabled();

        initMorningTime();
        initNoonTime();
        initEveningTime();
        initNightTime();
        return binding.getRoot();
    }

    private void initNoonTime() {
        String noonTime = getPrefs().getNoonTime();
        Date date = null;
        try {
            date = format.parse(noonTime);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        Calendar calendar = Calendar.getInstance();
        if (date != null) calendar.setTime(date);
        dayHour = calendar.get(Calendar.HOUR_OF_DAY);
        dayMinute = calendar.get(Calendar.MINUTE);
        dayTime.setText(TimeUtil.getTime(calendar.getTime(), is24));
    }

    private void initEveningTime() {
        String evening = getPrefs().getEveningTime();
        Date date = null;
        try {
            date = format.parse(evening);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        Calendar calendar = Calendar.getInstance();
        if (date != null) calendar.setTime(date);
        eveningHour = calendar.get(Calendar.HOUR_OF_DAY);
        eveningMinute = calendar.get(Calendar.MINUTE);
        eveningTime.setText(TimeUtil.getTime(calendar.getTime(), is24));
    }

    private void initNightTime() {
        String night = getPrefs().getNightTime();
        Date date = null;
        try {
            date = format.parse(night);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        if (date != null) calendar.setTime(date);
        nightHour = calendar.get(Calendar.HOUR_OF_DAY);
        nightMinute = calendar.get(Calendar.MINUTE);
        nightTime.setText(TimeUtil.getTime(calendar.getTime(), is24));
    }

    private void initMorningTime() {
        String morning = getPrefs().getMorningTime();
        Date date = null;
        try {
            date = format.parse(morning);
        } catch (ParseException e) {
            e.printStackTrace();
        }
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        if (date != null) calendar.setTime(date);
        morningHour = calendar.get(Calendar.HOUR_OF_DAY);
        morningMinute = calendar.get(Calendar.MINUTE);
        morningTime.setText(TimeUtil.getTime(calendar.getTime(), is24));
    }

    @Override
    public void onResume() {
        super.onResume();
        if (getCallback() != null) {
            getCallback().onTitleChange(getString(R.string.time));
            getCallback().onFragmentSelect(this);
        }
    }

    private void morningDialog() {
        TimeUtil.showTimePicker(getContext(), (view, hourOfDay, minute) -> {
            morningHour = hourOfDay;
            morningMinute = minute;
            Calendar calendar = Calendar.getInstance();
            calendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
            calendar.set(Calendar.MINUTE, minute);
            String time = format.format(calendar.getTime());
            getPrefs().setMorningTime(time);
            morningTime.setText(TimeUtil.getTime(calendar.getTime(), is24));
        }, morningHour, morningMinute);
    }

    private void dayDialog() {
        TimeUtil.showTimePicker(getContext(), (view, hourOfDay, minute) -> {
            dayHour = hourOfDay;
            dayMinute = minute;
            Calendar calendar = Calendar.getInstance();
            calendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
            calendar.set(Calendar.MINUTE, minute);
            String time = format.format(calendar.getTime());
            getPrefs().setNoonTime(time);
            dayTime.setText(TimeUtil.getTime(calendar.getTime(), is24));
        }, dayHour, dayMinute);
    }

    private void nightDialog() {
        TimeUtil.showTimePicker(getContext(), (view, hourOfDay, minute) -> {
            nightHour = hourOfDay;
            nightMinute = minute;
            Calendar calendar = Calendar.getInstance();
            calendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
            calendar.set(Calendar.MINUTE, minute);
            String time = format.format(calendar.getTime());
            getPrefs().setNightTime(time);
            nightTime.setText(TimeUtil.getTime(calendar.getTime(), is24));
        }, nightHour, nightMinute);
    }

    private void eveningDialog() {
        TimeUtil.showTimePicker(getContext(), (view, hourOfDay, minute) -> {
            eveningHour = hourOfDay;
            eveningMinute = minute;
            Calendar calendar = Calendar.getInstance();
            calendar.set(Calendar.HOUR_OF_DAY, hourOfDay);
            calendar.set(Calendar.MINUTE, minute);
            String time = format.format(calendar.getTime());
            getPrefs().setEveningTime(time);
            eveningTime.setText(TimeUtil.getTime(calendar.getTime(), is24));
        }, eveningHour, eveningMinute);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.morningTime:
                morningDialog();
                break;
            case R.id.dayTime:
                dayDialog();
                break;
            case R.id.eveningTime:
                eveningDialog();
                break;
            case R.id.nightTime:
                nightDialog();
                break;
        }
    }
}

File: app/src/main/java/com/elementary/tasks/navigation/settings/voice/TimeOfDayFragment.kt
package com.elementary.tasks.navigation.settings.voice

import android.os.Bundle
import android.view.View
import com.elementary.tasks.R
import com.elementary.tasks.core.utils.TimeUtil
import com.elementary.tasks.core.utils.toDate
import com.elementary.tasks.core.utils.toHm
import com.elementary.tasks.databinding.FragmentSettingsTimeOfDayBinding
import com.elementary.tasks.navigation.settings.BaseSettingsFragment
import org.koin.androidx.viewmodel.ext.android.viewModel
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

class TimeOfDayFragment : BaseSettingsFragment<FragmentSettingsTimeOfDayBinding>(), View.OnClickListener {

  private val viewModel by viewModel<TimesViewModel>()
  private var is24: Boolean = false
  private val format = SimpleDateFormat("HH:mm", Locale.getDefault())

  override fun layoutRes(): Int = R.layout.fragment_settings_time_of_day

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    binding.nightTime.setOnClickListener(this)
    binding.eveningTime.setOnClickListener(this)
    binding.dayTime.setOnClickListener(this)
    binding.morningTime.setOnClickListener(this)

    is24 = prefs.is24HourFormat

    initMorningTime()
    initNoonTime()
    initEveningTime()
    initNightTime()
  }

  override fun onStart() {
    super.onStart()
    viewModel.morningTime.observe(viewLifecycleOwner, {
      if (it != null) {
        binding.morningTime.text = TimeUtil.getTime(it.toDate(), is24, prefs.appLanguage)
      }
    })
    viewModel.dayTime.observe(viewLifecycleOwner, {
      if (it != null) {
        binding.dayTime.text = TimeUtil.getTime(it.toDate(), is24, prefs.appLanguage)
      }
    })
    viewModel.eveningTime.observe(viewLifecycleOwner, {
      if (it != null) {
        binding.eveningTime.text = TimeUtil.getTime(it.toDate(), is24, prefs.appLanguage)
      }
    })
    viewModel.nightTime.observe(viewLifecycleOwner, {
      if (it != null) {
        binding.nightTime.text = TimeUtil.getTime(it.toDate(), is24, prefs.appLanguage)
      }
    })
  }

  private fun initNoonTime() {
    val noonTime = prefs.noonTime
    var date: Date? = null
    try {
      date = format.parse(noonTime)
    } catch (e: ParseException) {
      e.printStackTrace()
    }
    viewModel.dayTime.postValue(date.toHm())
  }

  private fun initEveningTime() {
    val evening = prefs.eveningTime
    var date: Date? = null
    try {
      date = format.parse(evening)
    } catch (e: ParseException) {
      e.printStackTrace()
    }
    viewModel.eveningTime.postValue(date.toHm())
  }

  private fun initNightTime() {
    val night = prefs.nightTime
    var date: Date? = null
    try {
      date = format.parse(night)
    } catch (e: ParseException) {
      e.printStackTrace()
    }
    viewModel.nightTime.postValue(date.toHm())
  }

  private fun initMorningTime() {
    val morning = prefs.morningTime
    var date: Date? = null
    try {
      date = format.parse(morning)
    } catch (e: ParseException) {
      e.printStackTrace()
    }
    viewModel.morningTime.postValue(date.toHm())
  }

  override fun getTitle(): String = getString(R.string.time)

  private fun morningDialog() {
    viewModel.morningTime.value?.let { (hour, minute) ->
      timeDialog(hour, minute) { hourOfDay, minuteOfHour ->
        val hm = TimeUtil.HM(hour = hourOfDay, minute = minuteOfHour)
        prefs.morningTime = format.format(hm.toDate())
        viewModel.morningTime.postValue(hm)
      }
    }
  }

  private fun dayDialog() {
    viewModel.dayTime.value?.let { (hour, minute) ->
      timeDialog(hour, minute) { hourOfDay, minuteOfHour ->
        val hm = TimeUtil.HM(hour = hourOfDay, minute = minuteOfHour)
        prefs.noonTime = format.format(hm.toDate())
        viewModel.dayTime.postValue(hm)
      }
    }
  }

  private fun nightDialog() {
    viewModel.nightTime.value?.let { (hour, minute) ->
      timeDialog(hour, minute) { hourOfDay, minuteOfHour ->
        val hm = TimeUtil.HM(hour = hourOfDay, minute = minuteOfHour)
        prefs.nightTime = format.format(hm.toDate())
        viewModel.nightTime.postValue(hm)
      }
    }
  }

  private fun eveningDialog() {
    viewModel.eveningTime.value?.let { (hour, minute) ->
      timeDialog(hour, minute) { hourOfDay, minuteOfHour ->
        val hm = TimeUtil.HM(hour = hourOfDay, minute = minuteOfHour)
        prefs.eveningTime = format.format(hm.toDate())
        viewModel.eveningTime.postValue(hm)
      }
    }
  }

  private fun timeDialog(h: Int, m: Int, callback: (hourOfDay: Int, minuteOfHour: Int) -> Unit) {
    withContext {
      TimeUtil.showTimePicker(it, prefs.is24HourFormat, h, m) { _, hourOfDay, minute ->
        callback.invoke(hourOfDay, minute)
      }
    }
  }

  override fun onClick(v: View) {
    when (v.id) {
      R.id.morningTime -> morningDialog()
      R.id.dayTime -> dayDialog()
      R.id.eveningTime -> eveningDialog()
      R.id.nightTime -> nightDialog()
    }
  }
}
